examples/generative/bayesian_neural_networks_wine.py

```diff
@@ -0,0 +1,419 @@
+"""
+Title: Probabilistic Bayesian Neural Networks
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/15
+Last modified: 2021/01/15
+Description: Building probabilistic Bayesian neural network models with TensorFlow Probability.
+"""
+
+"""
+## Introduction
+
+Taking a probabilistic approach to deep learning allows to account for *uncertainty*,
+so that models can assign less levels of confidence to incorrect predictions.
+Sources of uncertainty can be found in the data, due to measurement error or
+noise in the labels, or the model, due to insufficient data availability for
+the model to learn effectively.
+
+
+This example demonstrates how to build basic probabilistic Bayesian neural networks
+to account for these two types of uncertainty.
+We use [TensorFlow Probability](https://www.tensorflow.org/probability) library,
+which is compatible with Keras API.
+
+This example requires TensorFlow 2.3 or higher.
+You can install Tensorflow Probability using the following command:
+
+```python
+pip install tensorflow-probability
+```
+"""
+
+"""
+## The dataset
+
+We use the [Wine Quality](https://archive.ics.uci.edu/ml/datasets/wine+quality)
+dataset, which is available in the [TensorFlow Datasets](https://www.tensorflow.org/datasets/catalog/wine_quality).
+We use the red wine subset, which contains 4,898 examples.
+The dataset has 11numerical physicochemical features of the wine, and the task
+is to predict the wine quality, which is a score between 0 and 10.
+In this example, we treat this as a regression task.
+
+You can install TensorFlow Datasets using the following command:
+
+```python
+pip install tensorflow-datasets
+```
+"""
+
+"""
+## Setup
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_datasets as tfds
+import tensorflow_probability as tfp
+
+"""
+## Create training and evaluation datasets
+
+Here, we load the `wine_quality` dataset using `tfds.load()`, and we convert
+the target feature to float. Then, we shuffle the dataset and split it into
+training and test sets. We take the first `train_size` examples as the train
+split, and the rest as the test split.
+"""
+
+
+def get_train_and_test_splits(train_size, batch_size=1):
+    dataset = tfds.load(name="wine_quality", as_supervised=True, split="train").map(
+        lambda x, y: (x, tf.cast(y, tf.float32))
+    )
+
+    train_dataset = (
+        dataset.take(train_size).shuffle(buffer_size=train_size).batch(batch_size)
+    )
+    test_dataset = dataset.skip(train_size).batch(batch_size)
+
+    return train_dataset, test_dataset
```

Comment:
Please add prefetching and caching. Please also add a comment stating that:

- We shuffle with a buffer the same size as the dataset because the dataset is very small and fits into memory.
- Same argument for caching.

Reply:
Done